src/engine.rs

```diff
@@ -1,17 +1,42 @@
 use borsh::{BorshDeserialize, BorshSerialize};
 use evm::backend::{Apply, ApplyBackend, Backend, Basic, Log};
-use evm::executor::{MemoryStackState, StackExecutor, StackSubstateMetadata};
+use evm::executor::{StackExecutor, StackSubstateMetadata};
 use evm::ExitFatal;
 use evm::{Config, CreateScheme, ExitError, ExitReason};
 
+use crate::contract::current_address;
 use crate::map::LookupMap;
-use crate::parameters::{FunctionCallArgs, NewCallArgs, SubmitResult, ViewCallArgs};
+use crate::parameters::{
+    FunctionCallArgs, NEP141FtOnTransferArgs, NewCallArgs, SubmitResult, ViewCallArgs,
+};
 use crate::precompiles;
-use crate::prelude::{Address, TryInto, Vec, H256, U256};
+use crate::prelude::{Address, ToString, TryInto, Vec, H256, U256};
 use crate::sdk;
+use crate::state::AuroraStackState;
 use crate::storage::{address_to_key, bytes_to_key, storage_to_key, KeyPrefix, KeyPrefixU8};
 use crate::types::{u256_to_arr, AccountId};
 
+macro_rules! unwrap_res_or_finish {
+    ($e:expr, $output:expr) => {
+        match $e {
+            Ok(v) => v,
+            Err(_) => {
+                sdk::return_output($output);
+                return;
+            }
+        }
+    };
+}
```

Comment:
I don't love this style of macro. It effectively does the same thing as the `?` operator, but more implicitly. It makes me think the need for this macro is a bit of a code smell that indicates some function should be returning a `Result` (and then could use the `?` operator), but isn't. If the function did return a result then the callsite would be the only place where this match is needed and so the macro would not be necessary.

Reply:
I tried to change the output to Result, but couldn't find a natural way to do it.

I tried using `Result<&[u8], &[u8]>`, but the output should be handled in the same way for `Ok` and `Err`. So handled the output with:

```
let output = match engine.receive_erc20_tokens(args) {
                Ok(output) => output,
                Err(output) => output,
};
```

The more complicated part is that the output is built inside the function, so we can't return &[u8], and should return Vec<u8>.

I will not handle this for now, and we can address it later to not block on this.

Reply:
Sure, we can address this separately, no need to block on it.

Returning `Result<&[u8], &[u8]>` seems odd though. I would have expected something like `Result<&[u8], Error>` and then at the callsite `call_to_fn(...).unwrap_or(default_value)`. If the value to return changes depending on where the error happened then `unwrap_or` doesn't work and it would have to be an explicit match where the different error cases map to different default values.

I haven't looked very closely at where this macro is used to know if there is a complexity I am missing. But again, I'm ok with not blocking the PR and coming back to this later.